src/rgw/rgw_auth_s3.cc

```diff
@@ -1269,21 +1479,132 @@ void AWSv4ComplMulti::modify_request_state(const DoutPrefixProvider* dpp, req_st
+
+    std::copy(parsing_buf.begin(), parsing_buf.begin() + lf_bytes,
+              trailer_vec.begin());
+    tbuf_pos += lf_bytes;
+    ttotal += lf_bytes;
+
+    while (ttotal < trailer_buf_size) {
+      const size_t received =
+          io_base_t::recv_body(trailer_vec.data() + tbuf_pos, trailer_buf_size);
+      dout(30) << "AWSv4ComplMulti: recv trailer received=" << received
+               << dendl;
+      if (received == 0) {
+        break;
+      }
+      tbuf_pos += received;
+      ttotal += received;
+    }
+
+    std::string_view expected_trailer_signature;
+    std::string calculated_trailer_signature;
+
+    if (ttotal > sarrlen("\r\n0;")) {
+      const std::string_view sv_trailer(trailer_vec.data() + sarrlen("\r\n0;"),
+                                        ttotal - sarrlen("\r\n0;"));
+
+      if (cct()->_conf->subsys.should_gather(ceph_subsys_rgw, 10))
+          [[unlikely]] {
+        ldout(cct(), 10) << "trailer_section: " << sv_trailer << dendl;
+      }
+
+      std::string_view mut_sv_trailer(sv_trailer);
+
+      auto chunk_signature =
+          mut_extract_helper<"chunk-signature=", "\r\n">(mut_sv_trailer);
+
+      std::string_view sig_sv;
+      if (get<0>(chunk_signature)) {
+        sig_sv = get<1>(chunk_signature);
+        sig_sv.remove_prefix(sarrlen("chunk-signature="));
+
+        ldout(cct(), 10) << "expected last chunk signature: " << sig_sv
+                         << " remaining: " << mut_sv_trailer << dendl;
+      }
+
+      trailer_map_t trailer_map;
+
+      if (x_amz_trailer) {
+        extract_trailing_headers(*x_amz_trailer, mut_sv_trailer, trailer_map);
+      }
+
+      auto trailer_signature =
+          mut_extract_helper<"x-amz-trailer-signature:", "\r\n">(
+              mut_sv_trailer);
+      if (get<0>(trailer_signature)) {
+        auto trailing_sig = get<1>(trailer_signature);
+        split_header(
+            trailing_sig,
+            [&](const std::string_view k, const std::string_view v) -> void {
+              expected_trailer_signature = v;
+              // propagate to req_info.env
+              put_prop(ys_header_mangle(k), v);
+            });
+
+        calculated_trailer_signature =
+            calc_v4_trailer_signature(trailer_map, final_chunk_signature);
+
+        ldout(cct(), 10) << "expected trailer signature="
+                         << expected_trailer_signature
+                         << "\n calculated trailer signature="
+                         << calculated_trailer_signature
+                         << "\n trailer bytes remaining/not consumed: "
+                         << mut_sv_trailer << dendl;
+      } /* matched trailer signature */
+    } /* have trailer */
+
+    if (expect_trailer_signature() &&
+        ((calculated_trailer_signature != expected_trailer_signature) &&
+         (!calculated_trailer_signature.empty()))) {
+      throw rgw::io::Exception(ERR_SIGNATURE_NO_MATCH, std::system_category());
```

Comment:
shouldn't we also fail if we expect a signature but don't find `x-amz-trailer-signature`? in that case `calculated_trailer_signature.empty()` would be true

Reply:
yes, I think that's correct